### ISSUE TITLE
[BUG]: Incorrect "Created on" date with Atom Feeds

### ISSUE THREAD

**Issue by sakkamade** [martinrotter/rssguard#416]
#### Brief description of the issue.

The column "Created on" displays incorrect date with Atom feed.
#### How to reproduce the bug?
1. Add some Atom feed
2. Download its messages which contain `<published>` and `<updated>` in single entry
3. Observe wrong date

#### What was the expected result?

The column "Created on" should display published date. 
Since the application fetches every entry of the feed only once (even if entry was updated), 
the "Created on" date which application shows is static, and I would like to see it to be correct.

#### What actually happened?

Instead of "Created on" it displays the date of "Updated on."

#### Other information

<details><summary>Example (click to show)</summary>

For instance, from such feed, "Created on" date of `<entry>` would be extracted  from `<updated>` block.

```
<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom" >
    <link href="https://example.com/feed.xml" rel="self" type="application/atom+xml" />
    <link href="https://example.com/" rel="alternate" type="text/html" />
    <updated>2020-12-11T01:29:18+00:00</updated>
    <id>https://example.com</id>
    <title type="html">Example dot com</title><subtitle>Example</subtitle>
    <entry>
        <title type="html">Example title</title>
        <link href="https://example.com/test1" rel="alternate" type="text/html" title="Example title" />
        <published>1999-01-24T04:51:14+00:00</published>
        <updated>2020-12-10T05:21:33+00:00</updated>
        <id>https://example.com/test1</id>
        <summary type="html">
        Example summary.
        </summary>
    </entry>
</entry>
</feed>
```
</details>

* OS: Arch Linux
* RSS Guard
Version: 3.9.2 (built on Linux/x86_64)
Build date: 4/17/21 8:31 AM
Qt: 5.15.2 (compiled against 5.15.2)

**Comment by martinrotter** [comment]
Hello @sakkamade I highly doubt that there is this particular bug in RSS Guard but I will test your example.

**Comment by martinrotter** [comment]
@sakkamade OK, I made the check. RSS Guard deliberately prefers `updated` because it gives more precise results than `published`. If RSS Guard used `published`, then some functionality would be lost, primarily on very "busy" feeds which often notify their users that something has changed via `updated`. There are specific use-cases for either case which work nicely with "updated" and also use-cases which would work good with "published" and it is quite hard to decide which date/time to use.

Also, look at [spec](https://datatracker.ietf.org/doc/html/rfc4287#section-4.2.15): `The "atom:updated" element is a Date construct indicating the most recent instant in time when an entry or feed was modified in a way the publisher considers significant.`

This is what majority of users want.

There could be several ways of fixing the problem and one way particularly handy - message filtering.

You can write a special "filter" in RSS Guard to automagically manipulate any data in incomming messages, including their date.

I wrote a simple filter for you.

```js
function filterMessage() {
  // Read raw contents of message and
  // convert to JSON.
  json = utils.fromXmlToJson(msg.rawContents);
  jsonObj = JSON.parse(json)
  
  // Read published date and parse it.
  publishedDate = jsonObj.entry.published.__text;
  parsedDate = new Date(publishedDate);

  // Set new date/time for message and
  // proceed.
  msg.created = parsedDate;
  return MessageObject.Accept;
}
```

![image](https://user-images.githubusercontent.com/1255302/119772177-11b46c80-bebf-11eb-8f27-8b29577c562d.png)

Simply create the filter, assign it to individual feeds and also remove existing messages from your feeds via "Cleanup database" feature and everything will start to work.

![image](https://user-images.githubusercontent.com/1255302/119772276-314b9500-bebf-11eb-8602-38bab17f172d.png)

> Since the application fetches every entry of the feed only once (even if entry was updated)

This is not true, RSS Guard fetches every entry available in the feed everytime and performs relatively robust decisions to decide if the entry is "updated" and must be overwritten in the DB.

RSS Guard detected `updated` tag correctly and subsequently decoded `2020-12-10T05:21:33+00:00` without problems as `1607577693000` miliseconds since Unix epoch, which is correct value as seen below.

![image](https://user-images.githubusercontent.com/1255302/119769381-a8caf580-beba-11eb-8275-c3ac8b184ed4.png)

This very value is also correctly stored in database.

![image](https://user-images.githubusercontent.com/1255302/119769410-b41e2100-beba-11eb-8b1c-ccaa797ae3cc.png)

RSS Guard displays this value correctly in **local** timezone, which is `+01:00` on my workstation.

![image](https://user-images.githubusercontent.com/1255302/119769492-d57f0d00-beba-11eb-9368-2c895a4ca8d5.png)

**Comment by sakkamade** [comment]
> > Since the application fetches every entry of the feed only once (even if entry was updated)
>
> This is not true, RSS Guard fetches every entry available in the feed everytime ...

Indeed, I apologise, I have tried too many RSS readers today and got mixed :heart: 

> and one way particularly handy - message filtering

So no more search for app and more documentation reading then. 
Thank you!

**Comment by martinrotter** [comment]
@sakkamade The message filtering feature is relatively new and some really useful sub-features might be missing. Let me know if that is the case.

**Comment by sakkamade** [comment]
>  Let me know if that is the case.

Naturally :heart:

**Comment by sakkamade** [comment]
That was too fast...
I guess I rather open a separate issue.

**Comment by martinrotter** [comment]
@sakkamade The message filtering feature is relatively new and some really useful sub-features might be missing. Let me know if that is the case. Also, note that the above filter will only work for `ATOM` feeds. `RSS` feeds will have line

`publishedDate = jsonObj.entry.published.__text;`

different.

Also, right now I pushed new code which adds more robust date/time conversion function into filtering mechanism. New version of your filter is now this:

```js
function filterMessage() {
  // Read raw contents of message and
  // convert to JSON.
  json = utils.fromXmlToJson(msg.rawContents);
  jsonObj = JSON.parse(json)
  
  // Read published date and parse it.
  publishedDate = jsonObj.entry.published.__text;
  parsedDate = utils.parseDateTime(publishedDate);

  // Set new date/time for message and
  // proceed.
  msg.created = parsedDate;
  return MessageObject.Accept;
}
```

It will only work with latest development build, because of line: `parsedDate = utils.parseDateTime(publishedDate);` which uses this backend [code](https://github.com/martinrotter/rssguard/blob/master/src/librssguard/miscellaneous/textfactory.cpp#L70) which is exactly the same code as used by RSS Guard when fetching messages.

https://github.com/martinrotter/rssguard/blob/master/resources/docs/Message-filters.md#utils-object

**Comment by sakkamade** [comment]
You should not remove your commentaries...

Another issue is not related to this, read here https://github.com/martinrotter/rssguard/issues/418.

**Comment by martinrotter** [comment]
Yes, I fucked up. :|

**Comment by sakkamade** [comment]
You should not be so hard on yourself...

**Comment by sakkamade** [comment]
@martinrotter, Might then "Created on" be renamed to just "Date"? To avod any confusion.

**Comment by martinrotter** [comment]
@sakkamade OK, renaming RN. I thought about having two columns "Created" "Updated" but that would make some other things more difficult.

**Comment by martinrotter** [comment]
done

**Comment by sakkamade** [comment]
> Also, look at spec: The "atom:updated" element is a Date construct indicating the most recent instant in time when an entry or feed was modified in a way the publisher considers significant.

I forgot to say, I am much obliged for this piece of intelligence in particular.

**Comment by sakkamade** [comment]
@martinrotter, You have forgotten to rename "Created on" in Message filter modal window.

**Comment by martinrotter** [comment]
done
